File: robot_attention_perception/async_fusion.py
from __future__ import annotations

from dataclasses import dataclass, field

from .fusion import AttentionFusion, FusionConfig
from .types import AcousticTrack, AttentionState, PerceptionFrame, RobotState, VisionPerson


@dataclass(frozen=True)
class AsyncFusionConfig:
    audio_ttl_ms: int = 180
    vision_ttl_ms: int = 450
    robot_ttl_ms: int = 250
    max_frame_rate_hz: float = 50.0


@dataclass
class AsyncPerceptionBuffer:
    """Fuse mixed-rate perception messages without waiting for exact timestamp matches."""

    fusion: AttentionFusion = field(default_factory=AttentionFusion)
    config: AsyncFusionConfig = field(default_factory=AsyncFusionConfig)
    acoustic_tracks: dict[str, AcousticTrack] = field(default_factory=dict)
    vision_people: dict[str, VisionPerson] = field(default_factory=dict)
    robot_state: RobotState = field(default_factory=lambda: RobotState(stamp_ms=0))
    last_update_ms: int = 0

    @classmethod
    def with_algorithm(
        cls,
        algorithm: str = "ros4hri_native",
        use_transcript_wake_words: bool = True,
    ) -> "AsyncPerceptionBuffer":
        return cls(
            fusion=AttentionFusion(
                FusionConfig(
                    algorithm=algorithm,
                    use_transcript_wake_words=use_transcript_wake_words,
                    # Without a wake word, a fresh final ASR transcript is
                    # the addressing signal. Use a slightly lower listen
                    # threshold so the transcript fallback (speech=1.0,
                    # clarity floor=0.7) can pass without fabricating a wake
                    # word bonus.
                    min_listen_score=0.52 if not use_transcript_wake_words else 0.62,
                )
            )
        )

    def ingest_acoustic(self, tracks: list[AcousticTrack] | tuple[AcousticTrack, ...]) -> None:
        for track in tracks:
            self.acoustic_tracks[track.track_id] = track

    def ingest_vision(self, people: list[VisionPerson] | tuple[VisionPerson, ...]) -> None:
        for person in people:
            self.vision_people[person.person_id] = person

    def ingest_robot(self, state: RobotState) -> None:
        self.robot_state = state

    def update(self, now_ms: int) -> AttentionState | None:
        min_period_ms = int(1000.0 / max(self.config.max_frame_rate_hz, 1.0))
        if self.last_update_ms and now_ms - self.last_update_ms < min_period_ms:
            return None
        self.last_update_ms = now_ms

        self._prune(now_ms)
        frame = PerceptionFrame(
            stamp_ms=now_ms,
            acoustic_tracks=[
                track.aged(now_ms - track.stamp_ms, self.config.audio_ttl_ms)
                for track in self.acoustic_tracks.values()
            ],
            vision_people=[
                person.aged(now_ms - person.stamp_ms, self.config.vision_ttl_ms)
                for person in self.vision_people.values()
            ],
            robot=(
                self.robot_state
                if now_ms - self.robot_state.stamp_ms <= self.config.robot_ttl_ms
                else RobotState(stamp_ms=now_ms)
            ),
        )
        state = self.fusion.update(frame)
        if any(now_ms - track.stamp_ms > 0 for track in self.acoustic_tracks.values()):
            state = self._append_reason(state, "async_audio_cache")
        if any(now_ms - person.stamp_ms > 0 for person in self.vision_people.values()):
            state = self._append_reason(state, "async_vision_cache")
        return state

    def _prune(self, now_ms: int) -> None:
        self.acoustic_tracks = {
            track_id: track
            for track_id, track in self.acoustic_tracks.items()
            if now_ms - track.stamp_ms <= self.config.audio_ttl_ms
        }
        self.vision_people = {
            person_id: person
            for person_id, person in self.vision_people.items()
            if now_ms - person.stamp_ms <= self.config.vision_ttl_ms
        }

    @staticmethod
    def _append_reason(state: AttentionState, reason: str) -> AttentionState:
        return AttentionState(
            stamp_ms=state.stamp_ms,
            target_id=state.target_id,
            target_kind=state.target_kind,
            confidence=state.confidence,
            listen=state.listen,
            addressed_to_robot=state.addressed_to_robot,
            source_track_id=state.source_track_id,
            person_id=state.person_id,
            azimuth_deg=state.azimuth_deg,
            reasons=tuple((*state.reasons, reason)),
            transcripts=state.transcripts,
        )

File: src/async_fusion.cpp
#include "robot_attention_perception/async_fusion.hpp"

#include <algorithm>
#include <cmath>
#include <utility>

namespace robot_attention_perception
{

float freshnessFactor(std::int64_t age_ms, std::int64_t max_age_ms)
{
  if (max_age_ms <= 0) {
    return 0.0F;
  }
  if (age_ms <= 0) {
    return 1.0F;
  }
  if (age_ms >= max_age_ms) {
    return 0.0F;
  }
  return 1.0F - (static_cast<float>(age_ms) / static_cast<float>(max_age_ms));
}

AcousticTrack AcousticTrack::aged(std::int64_t age_ms, std::int64_t max_age_ms) const
{
  if (age_ms <= 0) {
    return *this;
  }
  const float factor = freshnessFactor(age_ms, max_age_ms);
  AcousticTrack out = *this;
  out.voice_activity = out.voice_activity && factor > 0.0F;
  out.speech_probability *= factor;
  out.clarity *= factor;
  return out;
}

VisionPerson VisionPerson::aged(std::int64_t age_ms, std::int64_t max_age_ms) const
{
  if (age_ms <= 0) {
    return *this;
  }
  const float factor = freshnessFactor(age_ms, max_age_ms);
  VisionPerson out = *this;
  out.face_visible = out.face_visible && factor > 0.0F;
  out.face_confidence *= factor;
  out.gaze_score *= factor;
  out.body_facing_score *= factor;
  out.gesture_score *= factor;
  if (factor <= 0.25F) {
    out.engagement_status = "unknown";
  }
  if (factor <= 0.45F) {
    out.gesture.clear();
  }
  return out;
}

AsyncPerceptionBuffer::AsyncPerceptionBuffer(
  FusionConfig fusion_config, AsyncFusionConfig async_config)
: fusion_(std::move(fusion_config)), config_(async_config)
{
}

void AsyncPerceptionBuffer::ingestAcoustic(const std::vector<AcousticTrack> & tracks)
{
  for (const auto & track : tracks) {
    acoustic_tracks_[track.track_id] = track;
  }
}

void AsyncPerceptionBuffer::ingestVision(const std::vector<VisionPerson> & people)
{
  for (const auto & person : people) {
    vision_people_[person.person_id] = person;
  }
}

void AsyncPerceptionBuffer::ingestRobot(const RobotState & state)
{
  robot_state_ = state;
}

std::optional<AttentionState> AsyncPerceptionBuffer::update(std::int64_t now_ms)
{
  const auto min_period_ms = static_cast<std::int64_t>(
    std::floor(1000.0F / std::max(config_.max_frame_rate_hz, 1.0F)));
  if (last_update_ms_ != 0 && now_ms - last_update_ms_ < min_period_ms) {
    return std::nullopt;
  }
  last_update_ms_ = now_ms;
  prune(now_ms);

  PerceptionFrame frame;
  frame.stamp_ms = now_ms;
  for (const auto & item : acoustic_tracks_) {
    frame.acoustic_tracks.push_back(
      item.second.aged(now_ms - item.second.stamp_ms, config_.audio_ttl_ms));
  }
  for (const auto & item : vision_people_) {
    frame.vision_people.push_back(
      item.second.aged(now_ms - item.second.stamp_ms, config_.vision_ttl_ms));
  }
  if (now_ms - robot_state_.stamp_ms <= config_.robot_ttl_ms) {
    frame.robot = robot_state_;
  } else {
    frame.robot.stamp_ms = now_ms;
  }

  auto state = fusion_.update(frame);
  if (std::any_of(
      acoustic_tracks_.begin(), acoustic_tracks_.end(),
      [now_ms](const auto & item) { return now_ms - item.second.stamp_ms > 0; }))
  {
    appendReason(state, "async_audio_cache");
  }
  if (std::any_of(
      vision_people_.begin(), vision_people_.end(),
      [now_ms](const auto & item) { return now_ms - item.second.stamp_ms > 0; }))
  {
    appendReason(state, "async_vision_cache");
  }
  return state;
}

void AsyncPerceptionBuffer::prune(std::int64_t now_ms)
{
  for (auto it = acoustic_tracks_.begin(); it != acoustic_tracks_.end();) {
    if (now_ms - it->second.stamp_ms > config_.audio_ttl_ms) {
      it = acoustic_tracks_.erase(it);
    } else {
      ++it;
    }
  }
  for (auto it = vision_people_.begin(); it != vision_people_.end();) {
    if (now_ms - it->second.stamp_ms > config_.vision_ttl_ms) {
      it = vision_people_.erase(it);
    } else {
      ++it;
    }
  }
}

void AsyncPerceptionBuffer::appendReason(AttentionState & state, const std::string & reason)
{
  state.reasons.push_back(reason);
}

}  // namespace robot_attention_perception
